libflux/src/flux/formatter/tests.rs

```diff
@@ -0,0 +1,225 @@
+
+// This gives us a colorful diff.
+#[cfg(test)]
+use pretty_assertions::assert_eq;
+
+fn format_helper(golden: &str) {
+    let file = Parser::new(golden).parse_file("".to_string());
+    let mut fmt = Formatter::default();
+    fmt.format_file(&file, true);
+    let (ouput, _) = fmt.output();
+    assert_eq!(golden, ouput);
+}
+
+#[test]
+fn binary_op() {
+    format_helper("1 + 1 - 2");
+    format_helper("2 ^ 4");
+}
+
+#[test]
+fn funcs() {
+    format_helper("(r) =>\n\t(r.user == \"user1\")");
+    format_helper("add = (a, b) =>\n\t(a + b)"); // decl
+    format_helper("add(a: 1, b: 2)"); // call
+    format_helper("foo = (arg=[]) =>\n\t(1)"); // nil value as default
+    format_helper("foo = (arg=[1, 2]) =>\n\t(1)"); // none nil value as default
+}
+
+#[test]
+fn object() {
+    format_helper("{a: 1, b: {c: 11, d: 12}}");
+    format_helper("{foo with a: 1, b: {c: 11, d: 12}}"); // with
+    format_helper("{a, b, c}"); // implicit key object literal
+    format_helper("{\"a\": 1, \"b\": 2}"); // object with string literal keys
+    format_helper("{\"a\": 1, b: 2}"); // object with mixed keys
+}
+
+#[test]
+fn member() {
+    format_helper("object.property"); // member ident
+    format_helper("object[\"property\"]"); // member string literal
+}
+
+#[test]
+fn array() {
+    format_helper("a = [1, 2, 3]\n\na[i]");
+}
+
+#[test]
+fn conditional() {
+    format_helper("if a then b else c");
+    format_helper("if not a or b and c then 2 / (3 * 2) else obj.a(par: \"foo\")");
+}
+
+#[test]
+fn return_expr() {
+    format_helper("return 42");
+}
+
+#[test]
+fn option() {
+    format_helper("option foo = {a: 1}");
+    format_helper("option alert.state = \"Warning\""); // qualified
+}
+
+#[test]
+fn vars() {
+    format_helper("0.1"); // float
+    format_helper("100000000.0"); // integer float
+    format_helper("365d"); // duration
+    format_helper("1d1m1s"); // duration_multiple
+    format_helper("2018-05-22T19:53:00Z"); // time
+    format_helper("2018-05-22T19:53:01+07:00"); // zone
+    format_helper("2018-05-22T19:53:23.09012Z"); // nano sec
+    format_helper("2018-05-22T19:53:01.09012-07:00"); // nano with zone
+    format_helper("/^\\w+@[a-zA-Z_]+?\\.[a-zA-Z]{2,3}$/"); // regexp
+    format_helper("/^http:\\/\\/\\w+\\.com$/"); // regexp_escape
+}
+
+#[test]
+fn block() {
+    format_helper("foo = () => {\n\tfoo(f: 1)\n\t1 + 1\n}");
+}
+
+#[test]
+fn str_lit() {
+    format_helper(r#""foo""#);
+    format_helper(r#""this is\na string\nwith multiple lines""#); // multi lines
+    format_helper(r#""foo \\ \" \r\n""#); // with escape
+    format_helper(r#""\xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e""#); // with byte
+}
+
+#[test]
+fn package_import() {
+    format_helper("package foo\n"); // package
+    format_helper("import \"path/foo\"\nimport bar \"path/bar\""); // imports
+    format_helper(
+        "import foo \"path/foo\"\n\nfoo.from(bucket: \"testdb\")\n\
+         \t|> range(start: 2018-05-20T19:53:26Z)",
```

Comment:
Is there a way to write these tests without the `\t` and `\n` escape chars? That would make them easier to read.

Reply:
my editor has some issue with r#""#, let me try with other editors